modules/dnstap/dnstap.c

```diff
@@ -0,0 +1,410 @@
+	m.response_time_sec = now.tv_sec;
+	m.has_response_time_sec = true;
+	m.response_time_nsec = now.tv_sec * 1000;
+	m.has_response_time_nsec = true;
+
+	/* Qname */
+	m.query_zone.data = (uint8_t *)knot_pkt_qname(req->answer);
+	m.query_zone.len = req->answer->qname_size;
+	m.has_query_zone = true;
+
+	/* Create a dnstap Message */
+	Dnstap__Dnstap dnstap = DNSTAP__DNSTAP__INIT;
+	dnstap.type = DNSTAP__DNSTAP__TYPE__MESSAGE;
+	dnstap.message = (Dnstap__Message *)&m;
+
+	/* Pack the message */
+	uint8_t *frame = NULL;
+	size_t size = 0;
+	dt_pack(&dnstap, &frame, &size);
+	if (!frame) {
+		return kr_error(ENOMEM);
+	}
+
+	/* Submit a request to send message to fstrm_iothr*/
+	fstrm_res res = fstrm_iothr_submit(dnstap_dt->iothread, dnstap_dt->ioq, frame, size,
+			fstrm_free_wrapper, NULL);
+	if (res != fstrm_res_success) {
+		DEBUG_MSG("Error submitting dnstap message to iothr\n");
+		free(frame);
+		return kr_error(EBUSY);
+	}
+
+	return ctx->state;
+}
+
+KR_EXPORT
+int dnstap_init(struct kr_module *module) {
+	DEBUG_MSG("loading\n");
+	/* allocated memory for internal data */
+	struct dnstap_data *data = malloc(sizeof(*data));
+	if (!data) {
+		return kr_error(ENOMEM);
+	}
+	memset(data, 0, sizeof(*data));
+
+	/* save pointer to internal struct in module for future reference */
+	module->data = data;
+	return kr_ok();
+}
+
+KR_EXPORT
+int dnstap_deinit(struct kr_module *module) {
+	DEBUG_MSG("unloading\n");
+	struct dnstap_data *data = module->data;
+	/* Free allocated memory */
+	if (data) {
+		fstrm_iothr_destroy(&data->iothread);
+		DEBUG_MSG("fstrm iothread destroyed\n");
+		free(data);
+	}
+	return kr_ok();
+}
+
+/* dnstap_unix_writer returns a unix fstream writer
+ * https://gitlab.labs.nic.cz/labs/knot/blob/master/src/knot/modules/dnstap.c#L159
+ */
+static struct fstrm_writer* dnstap_unix_writer(const char *path) {
+	struct fstrm_unix_writer_options *opt = NULL;
+	struct fstrm_writer_options *wopt = NULL;
+	struct fstrm_writer *writer = NULL;
+
+	opt = fstrm_unix_writer_options_init();
+	if (!opt) {
+		return NULL;
+	}
+	fstrm_unix_writer_options_set_socket_path(opt, path);
+
+	wopt = fstrm_writer_options_init();
+	if (!wopt) {
+		fstrm_unix_writer_options_destroy(&opt);
+		return NULL;
+	}
+	fstrm_writer_options_add_content_type(wopt, DNSTAP_CONTENT_TYPE,
+			strlen(DNSTAP_CONTENT_TYPE));
+
+	writer = fstrm_unix_writer_init(opt, wopt);
+	fstrm_unix_writer_options_destroy(&opt);
+	fstrm_writer_options_destroy(&wopt);
+	if (!writer) {
+		return NULL;
+	}
+
+	fstrm_res res = fstrm_writer_open(writer);
+	if (res != fstrm_res_success) {
+		DEBUG_MSG("fstrm_writer_open returned %d\n", res);
+		fstrm_writer_destroy(&writer);
+		return NULL;
+	}
+
+	return writer;
+}
+
+/* find_string 
+ * create a new string from json
+ * *var is set to pointer of new string
+ * node must of type JSON_STRING
+ * new string can be at most len bytes
+ */
+static int find_string(const JsonNode *node, char **val, size_t len) {
+	if (!node || !node->key ) {
```

Comment:
Trailing whitespace after `->key` (in all json convenience functions below as well).